specs/ContextMenuRequested.md

```diff
@@ -0,0 +1,477 @@
+                        }
+                    }
+                    if(removeIndex > -1){
+                        CHECK_FAILURE(items->RemoveValueAtIndex(removeIndex));
+                    }
+                }
+                // Adding item
+                else if (selection_context == COREWEBVIEW2_SELECTION_CONTEXT_LINK)
+                {
+                    wil::com_ptr<ICoreWebView2Environment> webviewEnvironment;
+                    m_appWindow->GetWebViewEnvironment()->QueryInterface(
+                        IID_PPV_ARGS(&webviewEnvironment));
+                    wil::com_ptr<ICoreWebView2ContextMenuItem> newMenuItem;
+                    webviewEnvironment->CreateContextMenuItem(
+                        L"NewItem", L"Shorcut", &newMenuItem);    
+                    CHECK_FAILURE(items->AddValueAtIndex(newMenuItem, menuCollectionCount));
+                }
+                return S_OK;
+            })
+            .Get(),
+        &m_contextMenuRequestedToken));
+```
+
+## Win32 C++ Use Data to Display Custom Context Menu
+
+The developer can use the data provided in the Event arguments to display a custom context menu with entries of their choice. For this case, the developer specifies Handled to be true and requests a deferral. deferral of this event should be completed when the user selects a context menu item (either the app developer will handle the case, or can return the selected option to the browser) or when they click on the screen (effectively closing the menu).
+
+ ```cpp
+    webview2->add_ContextMenuRequested(
+        Callback<ICoreWebView2ContextMenuRequestedEventHandler>(
+            [this](
+                ICoreWebView2* sender,
+                ICoreWebView2ContextMenuRequestedEventArgs* args)
+            {
+                auto showMenu = [this, args] {
+                    wil::com_ptr<ICoreWebView2ContextMenuItemCollection> items;
+                    CHECK_FAILURE(args->get_ContextMenuItems(&items));
+                    args->put_Handled(true);
+
+                    UINT menuCollectionCount;
+                    CHECK_FAILURE(certificateCollection->get_Count(&menuCollectionCount));
+
+                    HMENU hPopupMenu = CreatePopupMenu();
+                    wil::com_ptr<ICoreWebView2ContextMenuItem> current;
+
+                    for(UINT i = 0; i < menuCollectionCount; i++) {
+                        CHECK_FAILURE(items->GetValueAtIndex(i, &current));
+                        wil::unique_cotaskmem_string name;
+                        COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR desc;
+                        wil::unique_cotaskmem_string shortcut;
+                        CHECK_FAILURE(current->get_Name(&name));
+                        CHECK_FAILURE(current->get_Descriptor(&desc));
+                        CHECK_FAILURE(current->get_Shortcut(&shortcut));
+                        InsertMenu(hPopupMenu, 0, MF_BYPOSITION | MF_STRING, desc, name + "-" + shortcut);
```

Comment:
Using 'desc' as the ID is a problem because the values aren't necessarily unique. The _CUSTOM value could come up many times from various extension menu items.

Reply:
Internally, each Context Menu entry has a unique id that is used for to execute the associated command. I changed the put_SelectedItem to use the context menu object instead of the descriptor. Each Context menu object will hold the unique id (private) to be used to invoke the command when selected.